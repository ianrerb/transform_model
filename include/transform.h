#ifndef TRANSFORM_H
#define TRANSFORM_H
#define PI M_PI 
#include"pricemodel.h"
#include<complex.h>
#include<fftw3.h>
#include<math.h>
#include<vector>
#include<iostream>
#include<string>

struct Option {double strike; double premium; }; //struct to hold strike, premium pairs

/********************************************
 * ABSTRACT BASE CLASS FOR TRANSFORM CLASSES 
 ********************************************/

class transform_base {//abc
  public:
    transform_base(unsigned int size, bool test = false):N_(size), testmode(test){ };
    virtual ~transform_base(){ };
    
    void N(const unsigned int &val){ N_ = val; };
    unsigned int N() const { return N_; };

    virtual std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const =0;

  protected:
    unsigned int N_;
   
    bool testmode;
    void printComplexArray(double complex array[], unsigned int length, std::string title) const {
      std::cout<<"Array: "<<title<<std::endl;
      for(unsigned int i=0; i!=length; i++){
	std::cout<<"("<<creal(array[i])<<","<<cimag(array[i])<<")"<<"  ";
	}
      std::cout<<std::endl;
      };
 };
 
/**********************************
 * BEGIN DEFINITION OF FFT CLASS 
 **********************************/

 class FFT : public transform_base {
  public:
    FFT(unsigned int size, double alpha_,double eta_, bool test = false) : transform_base(size,test), alpha(alpha_), eta(eta_){ }; 
    virtual ~FFT(){ }; 
    
    void Alpha(const double &val){ alpha = val; };
    void Eta(const double &val){eta = val; };
    
    double Alpha() const { return alpha; };
    double Eta() const { return eta; };
    virtual double Lambda() const {return 2.0*PI/(static_cast<double>(N_)*eta); };
    
    virtual std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const; //defined in FFTengine.cpp

  protected:
    double alpha, eta;

    void ComputeX(double complex X[], const double &Spot, const double &Strike, const double &C, const pricemodel &model) const;  //defined in FFTengine.cpp
  };

/**********************************
 * BEGIN DEFINITION OF FrFFT CLASS 
 **********************************/

class FrFFT : public FFT {
  public:
  FrFFT(unsigned int size, double alpha_, double eta_, double lambda_, bool test = false): FFT(size, alpha_, eta_,test), lambda(lambda_){};
  ~FrFFT(){ };
  
  void Lambda(const double &l){ lambda = l; };
  double Lambda() const { return lambda; };
  
  double Gamma() const { return eta*lambda/(2.0*PI); };

  std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const; //defined in FrFFT.cpp
  
  private:
    double lambda;
         
    void ComputeY(double complex Y[],const double complex X[]) const; //defined in FrFFT.cpp

    void ComputeZ(double complex Z[]) const; //defined in FrFFT.cpp

    void ComputeXi(double complex Xi[], const double complex Y[], const double complex Z[]) const; //defined in FrFFT.cpp

  };

#endif
