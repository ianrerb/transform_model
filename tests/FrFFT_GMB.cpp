#include"../include/transform.h"
#include<cmath>
#include<iostream>
#include<vector>

using namespace std;

int main(){
  double Spot, Strike, rate, q, T, vol;
  Spot = 100;
  rate = .01;
  q = 0.0;
  T = .1;
  vol = .25;

  GBM model(vol,rate,q,T);
  FrFFT Engine(0,1.5,.15,.01);

  vector<unsigned int> n; 
  vector<double> strikes; 
  
  strikes.push_back(80);
  strikes.push_back(100);
  strikes.push_back(120);

  n.push_back(16);
  n.push_back(64);
  n.push_back(256);

  for(vector<double>::iterator sit = strikes.begin(); sit!=strikes.end(); sit++){
    for(vector<unsigned int>::iterator nit = n.begin(); nit!=n.end(); nit++){
	  Engine.N(*nit);
	  cout<<Engine.Prices(Spot,*sit,exp(-rate*T),model)[*nit/2].premium<<"  ";
      }
    cout<<endl;
    }
  return 0;
  }
