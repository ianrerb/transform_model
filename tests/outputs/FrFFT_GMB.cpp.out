50958.5  205800  855333  
28.0507  181.327  4944.88  
0.0615145  0.64218  119.453  
0.000216693  0.00227433  2.8856  
0.000100256  8.05468e-06  0.0697072  
2.13328e-06  6.5063e-12  1.72878e-10  
0.0218807  8.79808e-11  3.76161e-09  
